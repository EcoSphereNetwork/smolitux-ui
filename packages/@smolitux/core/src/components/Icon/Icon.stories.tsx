import type { Meta, StoryObj } from '@storybook/react';
import { Icon, IconName } from './Icon';

const meta: Meta<typeof Icon> = {
  title: 'Components/Core/Icon',
  component: Icon,
  argTypes: {
    name: { control: 'text' },
    size: { control: 'text' },
    color: { control: 'color' },
    title: { control: 'text' },
    ariaHidden: { control: 'boolean' },
  },
  tags: ['autodocs'],
};

export default meta;

type Story = StoryObj<typeof Icon>;

export const Standard: Story = {
  args: { name: 'check' as IconName },
};

export const WithTitle: Story = {
  args: { name: 'info' as IconName, title: 'Information' },
};

export const Sizes: Story = {
  render: (args) => (
    <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
      <Icon {...args} size={16} />
      <Icon {...args} size={24} />
      <Icon {...args} size={32} />
    </div>
  ),
  args: { name: 'check' as IconName },
};

export const Colors: Story = {
  render: (args) => (
    <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
      <Icon {...args} color="green" />
      <Icon {...args} color="red" />
      <Icon {...args} color="blue" />
    </div>
  ),
  args: { name: 'alert' as IconName },
};
