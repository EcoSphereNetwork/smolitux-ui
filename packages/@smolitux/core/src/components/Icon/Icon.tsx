import React, { forwardRef, useId } from 'react';
import { cn } from '@smolitux/utils';

export const iconPaths = {
  check: <path strokeLinecap="round" strokeLinejoin="round" d="M5 13l4 4L19 7" />,
  info: (
    <path
      strokeLinecap="round"
      strokeLinejoin="round"
      d="M13 16h-1v-4h-1m1-4h.01M21 12a9 9 0 11-18 0 9 9 0 0118 0z"
    />
  ),
  alert: (
    <path
      strokeLinecap="round"
      strokeLinejoin="round"
      d="M12 9v2m0 4h.01M10.29 3.86L1.82 18a1 1 0 00.86 1.5h18.64a1 1 0 00.86-1.5L13.71 3.86a1 1 0 00-1.42 0z"
    />
  ),
  'arrow-left': (
    <path
      strokeLinecap="round"
      strokeLinejoin="round"
      d="M14 5l7 7m0 0l-7 7m7-7H3"
    />
  ),
} as const;

export type IconName = keyof typeof iconPaths;

export interface IconProps
  extends Omit<React.SVGProps<SVGSVGElement>, 'name' | 'color'> {
  /** Icon name */
  name: IconName;
  /** Icon size */
  size?: number | string;
  /** Stroke color */
  color?: string;
  /** Hide from accessibility tree */
  ariaHidden?: boolean;
  /** Accessible title */
  title?: string;
  /** ARIA role */
  role?: string;
}

export const Icon = forwardRef<SVGSVGElement, IconProps>(
  (
    {
      name,
      size = 24,
      color = 'currentColor',
      ariaHidden = true,
      title,
      role,
      className,
      ...rest
    },
    ref,
  ) => {
    const Path = iconPaths[name];
    const id = useId();

    if (!Path) {
      return <span ref={ref as any} {...rest} />;
    }

    const titleId = title ? `${name}-title-${id}` : undefined;

    return (
      <svg
        ref={ref}
        width={size}
        height={size}
        viewBox="0 0 24 24"
        fill="none"
        stroke={color}
        strokeWidth={2}
        aria-hidden={title ? undefined : ariaHidden}
        aria-labelledby={title ? titleId : undefined}
        role={role || (title ? 'img' : 'presentation')}
        className={cn(className)}
        data-icon={name}
        data-testid="icon"
        {...rest}
      >
        {title ? <title id={titleId}>{title}</title> : null}
        {Path}
      </svg>
    );
  },
);

Icon.displayName = 'Icon';

export default Icon;
