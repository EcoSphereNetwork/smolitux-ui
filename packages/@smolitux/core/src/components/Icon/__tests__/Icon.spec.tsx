import React from 'react';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Icon } from '../Icon';

describe('Icon', () => {
  test('snapshot', () => {
    const tree = renderer.create(<Icon name="check" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders title and aria attributes', () => {
    render(<Icon name="info" title="information" />);
    const svg = screen.getByRole('img');
    expect(svg).toHaveAttribute('aria-labelledby');
    expect(screen.getByText('information')).toBeInTheDocument();
  });

  test('hidden when no title', () => {
    render(<Icon name="alert" />);
    const svg = screen.getByTestId('icon');
    expect(svg).toHaveAttribute('aria-hidden', 'true');
  });
});
