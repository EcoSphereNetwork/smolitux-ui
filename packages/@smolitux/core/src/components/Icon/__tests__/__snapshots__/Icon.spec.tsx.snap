// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon snapshot 1`] = `
<svg
  aria-hidden={true}
  className=""
  data-icon="check"
  data-testid="icon"
  fill="none"
  height={24}
  role="presentation"
  stroke="currentColor"
  strokeWidth={2}
  viewBox="0 0 24 24"
  width={24}
>
  <path
    d="M5 13l4 4L19 7"
    strokeLinecap="round"
    strokeLinejoin="round"
  />
</svg>
`;
